joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForFunctionsCreator.scala

```diff
@@ -36,12 +36,21 @@ trait AstForFunctionsCreator(implicit withSchemaValidation: ValidationMode) { th
     * @return
     *   a method declaration with additional refs and types if specified.
     */
-  protected def astForMethodDeclaration(node: MethodDeclaration, isClosure: Boolean = false): Seq[Ast] = {
+  protected def astForMethodDeclaration(
+    node: MethodDeclaration,
+    isClosure: Boolean = false,
+    singletonMethodFullName: Option[String] = None
+  ): Seq[Ast] = {
     val isInTypeDecl  = scope.surroundingAstLabel.contains(NodeTypes.TYPE_DECL)
     val isConstructor = (node.methodName == Defines.Initialize) && isInTypeDecl
     val methodName    = node.methodName
     // TODO: body could be a try
-    val fullName = computeMethodFullName(methodName)
+
+    val fullName = singletonMethodFullName match {
```

Comment:
Could we not instead have `isSingletonMethod: Boolean = false` and if true, use `computeSingletonObjectMethodFullName` instead of `computeMethodFullName`? Would keep things consistent with the `isClosure` parameter

Reply:
I tried it this way originally, but it does complicate things since the base class that the singleton method is being defined on isn't defined on the `MethodDecl` that is passed into `astForMethodDeclaration`. Only wya I can think of off the top of my head to make this possible is to add a new type that MethodDecl and SingletonObjectMethodDecl both extend and use that for the `astForMethodDeclaration`

Reply:
I believe I've added a `ProcedureDeclaration` trait which might be usable here, or you can extend `MethodDeclaration`, either way this would be preferable to be handled in a similar way to the closures/lambas